sdk/core/azure-core/azure/core/polling/base_polling.py

```diff
@@ -0,0 +1,552 @@
+
+        for operation in self._lro_algorithms:
+            if operation.can_poll(initial_response):
+                self._operation = operation
+                break
+        else:
+            raise BadResponse("Unable to find status link for polling.")
+
+        try:
+            _raise_if_bad_http_status_and_method(self._initial_response.http_response)
+            self._status = self._operation.set_initial_status(initial_response)
+
+        except BadStatus as err:
+            self._status = "Failed"
+            raise HttpResponseError(response=initial_response.http_response, error=err)
+        except BadResponse as err:
+            self._status = "Failed"
+            raise HttpResponseError(
+                response=initial_response.http_response, message=str(err), error=err
+            )
+        except OperationFailed as err:
+            raise HttpResponseError(response=initial_response.http_response, error=err)
+
+    def run(self):
+        try:
+            self._poll()
+        except BadStatus as err:
+            self._status = "Failed"
+            raise HttpResponseError(
+                response=self._pipeline_response.http_response, error=err
+            )
+
+        except BadResponse as err:
+            self._status = "Failed"
+            raise HttpResponseError(
+                response=self._pipeline_response.http_response,
+                message=str(err),
+                error=err,
+            )
+
+        except OperationFailed as err:
+            raise HttpResponseError(
+                response=self._pipeline_response.http_response, error=err
+            )
+
+    def _poll(self):
+        """Poll status of operation so long as operation is incomplete and
+        we have an endpoint to query.
+
+        :param callable update_cmd: The function to call to retrieve the
+         latest status of the long running operation.
+        :raises: OperationFailed if operation status 'Failed' or 'Canceled'.
+        :raises: BadStatus if response status invalid.
+        :raises: BadResponse if response invalid.
+        """
+
+        while not self.finished():
+            self._delay()
+            self.update_status()
+
+        if failed(self.status()):
+            raise OperationFailed("Operation failed or canceled")
+
+        final_get_url = self._operation.get_final_get_url(self._pipeline_response)
+        if final_get_url:
+            self._pipeline_response = self.request_status(final_get_url)
+            _raise_if_bad_http_status_and_method(self._pipeline_response.http_response)
+
+    def _parse_resource(self, pipeline_response):
+        # type: (PipelineResponseType) -> Optional[Any]
+        """Assuming this response is a resource, use the deserialization callback to parse it.
+        If body is empty, assuming no resource to return.
+        """
+        response = pipeline_response.http_response
+        if not _is_empty(response):
```

Comment:
Do we really want to silently return `None` if the response is not valid json (e.g. for some reason, the service - or a proxy between the client and the service - returned HTML - which we've seen for some failure cases)?

Reply:
It's not what we do, we raise a DecodeError from _is_empty if it's not valid JSON.

Reply:
After the update based on your previous comment, `_is_empty` no longer raise, it will be the responsibility of `_deserialization_callback` to raise if body syntax is incorrect.

Reply:
Should it also be `_deserialization_callback`'s responsibility to return `None` on an empty response? I could see (at least in theory) that a fully headers/status-code-driven/custom LRO would continue to poll even with an empty response body. Or would we expect someone to override `_parse_response` altogether in that case?

Reply:
This method is only called when you call "result()", to get the final model. the usual implemetation is (from autorest):
```python
        def get_long_running_output(pipeline_response):
            deserialized = self._deserialize('Product', pipeline_response)

            if cls:
                return cls(pipeline_response, deserialized, {})
            return deserialized
```
This indeed means that if there was no body, you don't get `cls` called on your final result, so you can't sniff the headers (didn't realize that). 
So removing the `is_empty` is probably fair, to get `cls` even with LRO with no return schema

Reply:
fixed